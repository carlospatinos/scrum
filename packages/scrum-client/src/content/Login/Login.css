@media all and (min-width: 180px) {
  .Login {
    /* padding-top: 20px;
text-align: justify;
word-wrap: break-word; */
    /* background: linear-gradient(rgba(168, 183, 245, 0.5), transparent);
background-color: white; */
  }

  .Login form {
    margin: 0 auto;
    max-width: 320px;
  }

  /* Shared */
  .loginBtn {
    box-sizing: border-box;
    position: relative;
    /* width: 13em;  - apply for fixed size */
    margin: 0.2em;
    padding: 0 15px 0 46px;
    border: none;
    text-align: left;
    line-height: 34px;
    white-space: nowrap;
    border-radius: 0.2em;
    font-size: 16px;
    color: #FFF;
  }

  .loginBtn:before {
    content: "";
    box-sizing: border-box;
    position: absolute;
    top: 0;
    left: 0;
    width: 34px;
    height: 100%;
  }

  .loginBtn:focus {
    outline: none;
  }

  .loginBtn:active {
    box-shadow: inset 0 0 0 32px rgba(0, 0, 0, 0.1);
  }


  /* Facebook */
  .loginBtn--facebook {
    background-color: #4C69BA;
    background-image: linear-gradient(#4C69BA, #3B55A0);
    /*font-family: "Helvetica neue", Helvetica Neue, Helvetica, Arial, sans-serif;*/
    text-shadow: 0 -1px 0 #354C8C;
  }

  .loginBtn--facebook:before {
    border-right: #364e92 1px solid;
    background: url('https://s3-us-west-2.amazonaws.com/s.cdpn.io/14082/icon_facebook.png') 6px 6px no-repeat;
  }

  .loginBtn--facebook:hover,
  .loginBtn--facebook:focus {
    background-color: #5B7BD5;
    background-image: linear-gradient(#5B7BD5, #4864B1);
  }


  /* Google */
  .loginBtn--google {
    /*font-family: "Roboto", Roboto, arial, sans-serif;*/
    background: #DD4B39;
    /* background-color: red; */
  }

  .loginBtn--google:before {
    border-right: #E74B37 1px solid;
    background: url('https://s3-us-west-2.amazonaws.com/s.cdpn.io/14082/icon_google.png') 6px 6px no-repeat;
  }

  .loginBtn--google:hover,
  .loginBtn--google:focus {
    background: #BB3F30;
  }

  /* Twitter */
  .loginBtn--twitter {
    /*font-family: "Roboto", Roboto, arial, sans-serif;*/
    background: #1bb2e9;
  }

  .loginBtn--twitter:before {
    border-right: #1B8EB8 1px solid;
    background: url('https://i.imgur.com/sO5Y2aG.png') 6px 6px no-repeat;
  }

  .loginBtn--twitter:hover,
  .loginBtn--twitter:focus {
    background: #1B8EB8;
  }

  .separator {
    display: flex;
    align-items: center;
    text-align: center;
  }

  .separator::before,
  .separator::after {
    content: '';
    flex: 1;
    border-bottom: 1px solid #000;
  }

  .separator::before {
    margin-right: .25em;
  }

  .separator::after {
    margin-left: .25em;
  }
}