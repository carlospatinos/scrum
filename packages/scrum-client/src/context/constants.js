const ACTIONS_FETCH = 'FETCH';
const ACTIONS_POST = 'POST';
const ACTIONS_FAILURE = 'FAILURE';
const ACTIONS_SUCCESS = 'SUCCESS';
const LIBRARY_ROOT = 'SCRUM';
const ROUTE_CHANGED = 'ROUTE_CHANGED';

export {
  ACTIONS_FETCH,
  ACTIONS_POST,
  ACTIONS_FAILURE,
  ACTIONS_SUCCESS,
  LIBRARY_ROOT,
  ROUTE_CHANGED,
};
