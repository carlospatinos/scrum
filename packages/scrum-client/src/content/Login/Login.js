import React, { useState } from "react";
import {
  Alert,
  Button,
  FormGroup,
  FormControl,
  FormLabel,
} from "react-bootstrap";
import { useHistory } from 'react-router-dom';

import { API_BASE_URL, ACCESS_TOKEN_NAME } from '../../constants/apiConstants';

import "./Login.css";

// import TagManager from 'react-gtm-module'

// if (process.env.NODE_ENV === "development" && !!process.env.REACT_APP_GTM_ID) {
//   document.title = "login";
//   const tagManagerArgs = {
//     dataLayer: {
//       page: "login", //Specific to each page
//       pagePath: window.location.pathname + window.location.search, //"/login", //Specific to each page
//       title: "login"
//     },
//     dataLayerName: "PageDataLayer"
//   };
//   TagManager.dataLayer(tagManagerArgs);
// }

export default function Login() {

  const history = useHistory();

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [apiResponse, setApiResponse] = useState("");
  function validateForm() {
    return email.length > 0 && password.length > 0;

  }

  function handleSubmit(event) {
    event.preventDefault();

    const requestOptions = {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password }),
    };

    try {
      fetch(API_BASE_URL + "/api/login", requestOptions)
        .then((response) => response.json())
        .then((data) => {
          console.log(data);
          if (data.isAuth) {
           localStorage.setItem(ACCESS_TOKEN_NAME, data.token);
            history.push("/home");
          } else {
            setApiResponse(data.message);
          }
        });
    } catch (e) {
      // console.log(e);
    }
  }
  const isValidForm = validateForm();

  return (
    <div className="Login">
      <form onSubmit={handleSubmit}>
        <FormGroup controlId="email">
          <FormLabel>Email</FormLabel>
          <FormControl
            autoFocus
            type="email"
            value={email}
            onChange={e => setEmail(e.target.value)}
          />
        </FormGroup>
        <FormGroup controlId="password">
          <FormLabel>Password</FormLabel>
          <FormControl
            value={password}
            onChange={e => setPassword(e.target.value)}
            type="password"
          />
        </FormGroup>
        {apiResponse && <Alert variant="danger">{apiResponse}</Alert>}
        <Button
          block
          disabled={!isValidForm}
          variant={isValidForm ? 'primary' : 'secondary'}
          type="submit"
        >
          Login
        </Button>
      </form>
    </div>
  );
}
