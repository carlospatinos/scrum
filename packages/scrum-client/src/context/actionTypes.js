import ContextUtil from './utils';
import { ACTIONS_POST } from './constants';

const LOGIN_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGIN',
  restVerb: ACTIONS_POST,
});
const LOGOUT_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGOUT',
});

export { LOGIN_ACTIONS, LOGOUT_ACTIONS };
