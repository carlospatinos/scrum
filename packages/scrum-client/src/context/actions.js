/* eslint-disable no-console */
import { END_POINTS } from 'scrum-common';
import { API_CONSTANTS } from '../constants';

export async function loginUser(dispatch, loginPayload) {
  const requestOptions = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    body: JSON.stringify(loginPayload),
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.AUTH_LOCAL}`,
      requestOptions
    );
    const data = await response.json();

    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function login3ppUser(dispatch) {
  const requestOptions = {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    credentials: 'include',
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.LOGIN_SUCCESS}`,
      requestOptions
    );

    const data = await response.json();
    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
    console.log(data.message);
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function logout(dispatch) {
  dispatch({ type: 'LOGOUT' });
  localStorage.removeItem(API_CONSTANTS.CURRENT_USER);
  localStorage.removeItem(API_CONSTANTS.ACCESS_TOKEN_NAME);
  // TODO is this needed? localStorage.clear();
  localStorage.clear();
}
