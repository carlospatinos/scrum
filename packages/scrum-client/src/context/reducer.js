/* eslint-disable no-shadow */
import { CURRENT_USER } from '../constants/apiConstants';

const user = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).user
  : '';
const token = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).login_access_token
  : '';

export const initialState = {
  user: '' || user,
  login_access_token: '' || token,
  loading: false,
  errorMessage: null,
};

export const AuthReducer = (initialState, action) => {
  switch (action.type) {
    case 'REQUEST_LOGIN':
      return {
        ...initialState,
        loading: true,
      };
    case 'LOGIN_SUCCESS':
      return {
        ...initialState,
        user: action.payload.user,
        login_access_token: action.payload.login_access_token,
        loading: false,
      };
    case 'LOGOUT':
      return {
        ...initialState,
        user: '',
        login_access_token: '',
      };

    case 'LOGIN_ERROR':
      return {
        ...initialState,
        loading: false,
        errorMessage: action.error,
      };

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};
