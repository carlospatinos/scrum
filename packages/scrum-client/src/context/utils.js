import { ACTIONS_FETCH, ACTIONS_SUCCESS, ACTIONS_FAILURE, LIBRARY_ROOT } from './constants';
/**
 * Generates an action type by concatenating the library, feature name and key.
 *
 * @param {String} library
 * @param {String} featureName
 * @param {String} key
 */
const generateActionType = (library, featureName, key) => `${library}/${featureName}/${key}`;

/**
 * Generates a partial implementation of `generateActionType` fixing library and featureName.
 *
 * @param {String} library
 * @param {String} featureName
 */
const generateFeatureActionType = (library, featureName) => key =>
  generateActionType(library, featureName, key);

/**
 * Generates the three default action types, baseAction,  successAction and failureAction of a given feature.
 *
 * @param {*} { library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }
 * @returns
 */
const generateActionTypes = ({ library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }) => {
  const baseActionKey = `${restVerb}_${featureName.toUpperCase()}`;
  const featureActionType = generateFeatureActionType(library, featureName);
  const baseAction = featureActionType(baseActionKey);
  const successAction = featureActionType(`${baseActionKey}_${ACTIONS_SUCCESS}`);
  const failureAction = featureActionType(`${baseActionKey}_${ACTIONS_FAILURE}`);
  return { baseAction, successAction, failureAction };
};
/**
 * Generates a context action, it handles the three action types of a given feature action.
 * - dispatchFxn( type: action.baseAction): Before invoking the `apiFxn`/
 * - dispatchFxn( type: action.successAction): When the response of  `apiFxn` is received.
 * - dispatchFxn( type: action.baseAction):  When an error  of `apiFxn` is received.
 *
 * @param {Function} dispatchFxn - The dispatcher function.
 * @param {Function} apiFxn - The API function to invoke.
 * @param {Object} action - The baseAction, successAction and failureAction of a given feature.
 */
const generateAction = (dispatchFxn, apiFxn, action) => async req => {
  const initialAction = {
    type: action.baseAction,
    payload: {},
  };
  dispatchFxn(initialAction);

  try {
    const response = await apiFxn(req);
    console.log('response.data', response.data);
    dispatchFxn({
      type: action.successAction,
      payload: response.data,
    });
    return response;
  } catch (e) {
    dispatchFxn({
      type: action.failureAction,
      error: e.message,
    });
    return undefined;
  }
};

/**
 * Context utility , it generates actions types and actions.
 */
const ContextUtil = { generateActionType, generateActionTypes, generateAction };
export default ContextUtil;
