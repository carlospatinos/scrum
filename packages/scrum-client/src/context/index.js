import { loginUser, login3ppUser, logout } from './actions';
import { AuthProvider, useAuthDispatch, useAuthState } from './context';

export { AuthProvider, useAuthState, useAuthDispatch, loginUser, login3ppUser, logout };
