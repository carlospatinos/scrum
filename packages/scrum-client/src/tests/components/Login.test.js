import { render, cleanup } from '../test-utils';
import Login from '../../content/Login';
import { useAuthState } from '../../context/context';
import { basicUser } from '../mocks/user';

jest.mock('../../context/context');
afterEach(cleanup);

describe('Login', () => {
  test('To match snapshot', () => {
    //   mock hook call
    useAuthState.mockReturnValue(basicUser);

    // render the component
    const { asFragment } = render(<Login />);

    // validates vs snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
