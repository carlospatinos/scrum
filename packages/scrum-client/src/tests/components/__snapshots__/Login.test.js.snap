// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login To match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-3 shadow-lg p-3 mb-5 bg-white rounded jumbotron"
  >
    <form
      class=""
    >
      <h1
        class="text-center"
      >
        Sign in
      </h1>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="email"
          placeholder="Enter email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="password"
          placeholder="Enter password"
          type="password"
          value=""
        />
        <div
          class="form-row"
        >
          <div
            class="forgot-password text-right col"
          >
            <a
              href="/forgot"
            >
              <small>
                Forgot your password?
              </small>
            </a>
          </div>
        </div>
      </div>
      <button
        class="btn btn-secondary btn-block"
        disabled=""
        type="submit"
      >
        Login
      </button>
      <div
        class="form-row"
      >
        <div
          class="forgot-password text-right col"
        >
          <a
            href="/signup/:referrer?"
          >
            <small>
              Do not have an account? Sing Up
            </small>
          </a>
        </div>
      </div>
      <br />
      <div
        class="separator"
      >
         OR 
      </div>
      <br />
      <button
        class="loginBtn loginBtn--google btn-block btn btn-primary"
        type="button"
      >
        Login with Google
      </button>
      <button
        class="loginBtn loginBtn--twitter btn-block btn btn-primary"
        type="button"
      >
        Login with Twitter
      </button>
      <button
        class="loginBtn loginBtn--facebook btn-block btn btn-primary"
        type="button"
      >
        Login with Facebook
      </button>
      <br />
    </form>
  </div>
</DocumentFragment>
`;
